### PR TITLE
[ProgressView] Allow the progress view to be sized to fractional points.

### LINKED ISSUE #9439: [ProgressView] Internal issue: b/147684724
This was filed as an internal issue. If you are a Googler, please visit [b/147684724](http://b/147684724) for more details.

---
#### Internal data
- Associated internal bug: [b/147684724](http://b/147684724)

### PR DESCRIPTION
**Pull request by jverkoey**
This allows the progress view to more smoothly represent fractional progress and to take advantage of multi-pixel points on devices with retina displays.

Closes https://github.com/material-components/material-components-ios/issues/9439#event-2955279250

### CODE REVIEW

File: components/ProgressView/src/MDCProgressView.m [hunk 1]
```diff
@@ -292,7 +292,7 @@ + (UIColor *)defaultTrackTintColorForProgressTintColor:(UIColor *)progressTintCo
 
 - (void)updateProgressView {
   // Update progressView with the current progress value.
-  CGFloat progressWidth = MDCCeil(self.progress * CGRectGetWidth(self.bounds));
```
Comment (romoore) [inline components/ProgressView/src/MDCProgressView.m:295]: Tip: Please use `MDCCeilScaled` which will align the width to pixel values of the screen but keep the "ceiling" behavior.  I believe that aligns with the goal of this PR and will prevent alignment on sub-pixel boundaries, which results in blurrier/less-crisp rendering.

**Review by romoore — APPROVED** [review]
LGTM but would recommend switching to a scale-based implementation rather than dropping the ceiling function entirely.

File: components/ProgressView/src/MDCProgressView.m (hunk 1 above)
Comment (jverkoey) [reply components/ProgressView/src/MDCProgressView.m:295]: Done.

File: components/ProgressView/src/MDCProgressView.m [hunk 2]
```diff
@@ -292,7 +292,9 @@ + (UIColor *)defaultTrackTintColorForProgressTintColor:(UIColor *)progressTintCo
...
+  CGFloat scale = self.window.screen ? self.window.screen.scale : 1;
```
Comment (romoore) [inline components/ProgressView/src/MDCProgressView.m]: Tip: Treating a CGFloat as a Boolean seems pretty risky. 

Suggestion: 

```objc
CGFloat scale = self.window.screen.scale > 0 ? self.window.screen.scale : 1;
```

File: components/ProgressView/src/MDCProgressView.m [hunk 3]
```diff
@@ -292,7 +292,9 @@ + (UIColor *)defaultTrackTintColorForProgressTintColor:(UIColor *)progressTintCo
...
+  CGFloat width = self.progress * CGRectGetWidth(self.bounds);
+  CGFloat progressWidth = MDCRound(width * scale) / scale;
```
Comment (romoore) [inline components/ProgressView/src/MDCProgressView.m]: Did you accidentally switch from ceiling to rounding behavior?  MDCProgressView has been using ceiling functions since 2016 and I think switching should be part of a different PR.  I think it's best to keep this one focused on the pixel-alignment issue.

I think `MDCCeilScaled` would be helpful here, but you could use `MDCCeil` if you don't like the scaled implementation.

**Review by romoore — CHANGES_REQUESTED** [review]
This PR might have accidentally become a functional change (ceil/round) rather than an alignment "clean up".

File: components/ProgressView/src/MDCProgressView.m (hunk 3 above)
Comment (jverkoey) [reply components/ProgressView/src/MDCProgressView.m]: I did change it intentionally in an effort to fix some surprising behavior, but I agree that we can approach that separately so for now I've added some unit tests to demonstrate the surprising behavior I encountered.

Comment (jverkoey) [reply components/ProgressView/src/MDCProgressView.m]: *and reverted this back to ceil for now.

Comment (jverkoey) [reply components/ProgressView/src/MDCProgressView.m]: Note the new `test97` tests in the new unit tests and the fact that the frame width is actually 98 on 1pt displays.

File: components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+- (void)test65 {
+  // When
+  self.progressView.progress = (float)0.65;
+  [self.progressView layoutIfNeeded];
+
+  // Then
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.origin.x, 0, 0.001);
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.origin.y, 0, 0.001);
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.size.width, 65, 0.001);
```
Comment (romoore) [inline components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m]: Tip: Please use `CGRectGetXXX()` methods for accessing the frame/bounds.  You could also standardize the rect first if you like.  Making this optional since it's test code, but it is considered a team norm.

File: components/ProgressView/tests/unit/ProgressViewProgressWithoutScreenTests.m [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  // When
+  self.progressView.progress = (float)0.97;
+  [self.progressView layoutIfNeeded];
+
+  // Then
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.origin.x, 0, 0.001);
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.origin.y, 0, 0.001);
+  // Note the surprising value of 98 here; this is due to the use of MDCCeil to calculate
+  // screen-aligned pixel values for the progress view's frame.
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.size.width, 98, 0.001);
```
Comment (romoore) [inline components/ProgressView/tests/unit/ProgressViewProgressWithoutScreenTests.m]: Tip: If these are the existing behavior, it may be worth landing the tests before the component update.  That would demonstrate that the refactor has no effect on non-retina screens but does produce improvements to retina displays.

File: components/ProgressView/src/MDCProgressView.m (hunk 3 above)
Comment (jverkoey) [reply components/ProgressView/src/MDCProgressView.m]: Filed https://github.com/material-components/material-components-ios/issues/9482 to track.

File: components/ProgressView/src/MDCProgressView.m (hunk 2 above)
Comment (jverkoey) [reply components/ProgressView/src/MDCProgressView.m]: Done, though to clarify: I was not using CGFloat as a Boolean here; I was testing if the screen is nil before making use of it. The > 0 check on the scale is safer in case the screen scale is somehow 0, so I've switched to that.

**Comment by jverkoey** [comment]
Thank you for the reviews Rob :)

File: components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.origin.x, 0, 0.001);
+  XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.origin.y, 0, 0.001);
+  CGFloat screenScale = self.progressView.window.screen.scale;
+
+  if (MDCCGFloatEqual(screenScale, 1)) {
+    // Note the surprising value of 98 here; this is due to the use of MDCCeil to calculate
+    // screen-aligned pixel values for the progress view's frame.
+    // TODO(https://github.com/material-components/material-components-ios/issues/9482): This should
+    // be 97.
+    XCTAssertEqualWithAccuracy(self.progressView.progressView.frame.size.width, 98, 0.001);
```
Comment (romoore) [inline components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m]: This is 97 when I test it on an iPad 2 and iPhone 4s, but 98 on 64-bit devices. Given its fragile nature, I'm not sure if it's worth keeping or not.  What do you think?

**Review by romoore — APPROVED** [review]
LGTM, with one concern about some fragile tests.  The rest are recommendations.

File: components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m (hunk 6 above)
Comment (jverkoey) [reply components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m]: Also looks like our test runners aren’t covering this case unfortunately.

Comment (romoore) [reply components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m]: They're not, but this is the type of scenario where I believe we still rely on maintainers/authors to conduct sufficient testing.

Comment (romoore) [reply components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m]: Looks like we could add 32-bit simulator testing if https://github.com/google/xctestrunner/issues/14 is fixed.

File: components/ProgressView/src/MDCProgressView.m (hunk 2 above)
Comment (romoore) [reply components/ProgressView/src/MDCProgressView.m]: Ah, yes. I clearly need to increase my coffee dosage...  ☕️ 📈

File: components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m (hunk 6 above)
Comment (jverkoey) [reply components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m]: > They're not, but this is the type of scenario where I believe we still rely on maintainers/authors to conduct sufficient testing.

+1, but I'm unfortunately not aware of a way to do so from Catalina without having a 32 bit device on hand and our test infra should be giving us comprehensive enough coverage for behavior deltas like this.

All that being said, I'm going to move these unit tests out to a separate PR so we can discuss further there.

Comment (romoore) [reply components/ProgressView/tests/unit/ProgressViewProgressWithCurrentScreenScaleTests.m]: Catalina is a problem, but right now I believe [our supported Xcode version is Xcode 10](https://github.com/material-components/material-components-ios/tree/develop/docs/build-env#xcode).  I think we should try to identify a way to get 32-bit unit tests running in CI until we drop iOS 9.

**Comment by jverkoey** [comment]
I've removed the unit tests and will add them in a separate PR.

I've confirmed that adding an iPhone 4s simulator to the kokoro job causes the unit tests to fail as described by Rob, so I'll be adding those tests independently.

**Comment by jverkoey** [comment]
This was rolled back as part of https://github.com/material-components/material-components-ios/pull/9494

**Comment by romoore** [comment]
Any background or reasons why it was rolled back? Did it have a bug?

**Comment by jverkoey** [comment]
This was rolled back because it was generating a non-trivial number of internal screenshot test failures and creating a risk for the release to land today.

**Comment by romoore** [comment]
Awesome. Glad to know it was a good change, but required lots of screenshot updates. :)
